dev/buildtool/build_commands.py

```diff
@@ -0,0 +1,732 @@
+        ' gradle hanging.')
+    self.add_argument(
+        parser, 'max_local_builds', defaults, None, type=int,
+        help='Maximum build concurrency.')
+    self.add_argument(
+        parser, 'run_unit_tests', defaults, False, action='store_true',
+        help='Run unit tests during build for all components other than Deck.')
+    self.add_argument(
+        parser, 'build_jar_repository', defaults, None,
+        help='bintray repository to publish jar files into.')
+    self.add_argument(
+        parser, 'maven_custom_init_file', defaults,
+        os.path.join(os.path.dirname(__file__), '..', 'maven-init.gradle'),
+        help='Path to a gradle init file to add to the debian builds.'
+        'Used to specify any custom behavior in the gradle builds.'
+        'Argument is a file path relative to the directory this script is'
+        ' executed in.'
+        ' The default value assumes we run this script from the parent'
+        ' directory of spinnaker/spinnaker.')
+    self.add_argument(
+        parser, 'force_clean_gradle_cache', defaults, False,
+        help='Force a fresh new component-specific gradle cache for'
+        ' each component build')
+    self.add_argument(
+        parser, 'gradle_cache_path', defaults,
+        '{home}/.gradle'.format(home=os.environ['HOME'])
+        if os.environ.get('HOME') else None,
+        help='Path to a gradle cache directory to use for the builds.')
+    self.add_argument(
+        parser, 'extra_gradle_args', defaults, None,
+        help='Additional arguments to pass to gradle (comma-delimited).')
+
+
+class GradleCommandProcessor(RepositoryCommandProcessor):
+  """Base class for build commands using Gradle."""
+  # pylint: disable=abstract-method
+
+  def __init__(self, factory, options, upstream_boms, **kwargs):
+    self.__upstream_boms = upstream_boms
+    self.__nebula_repository_version = {}
+
+    # This is to cut down on logging noise so we only emmit one
+    # message on the test against freshness rather than on every access.
+    self.__logged_is_newer = {}
+    super(GradleCommandProcessor, self).__init__(
+        factory, options, use_threadpool=options.use_threadpool, **kwargs)
+
+  def _get_nebula_repository_version(self, name):
+    """Get the GitSummary for the named repo."""
+    if not self.__nebula_repository_version.get(name):
+      summary_path = os.path.join(
+          self.options.scratch_dir, name,
+          '{name}-summary.yml'.format(name=name))
+      if not os.path.exists(summary_path):
+        logging.debug('Forcing a repository update because there is no %s',
+                      summary_path)
+        self._get_nebula_repository_path(name, force_new=True)
+
+      with open(summary_path, 'r') as stream:
+        data = yaml.load(stream)
+        self.__nebula_repository_version[name] = data['version']
+    return self.__nebula_repository_version[name]
+
+  def _do_determine_source_repositories(self):
```

Comment:
Generally single and double underscore functions are meant to be private, seems a little strange to be overriding. I missed this in the interface definition. I think the underscore is signalling this is used in another public function in the superclass, which is OK.

Reply:
Relevant SO: https://stackoverflow.com/questions/1301346/what-is-the-meaning-of-a-single-and-a-double-underscore-before-an-object-name - I think this is fine but just be aware of the limitations, e.g. `from M import *` won't import single underscore functions.

Reply:
The single underscores are intended to be protected. The methods starting with _do_ are meant to be hook methods that derived classes can specialize. I'm using inheritance here so always super. This usage is entirely consistent with that thread.

These are generally methods, not functions. Top level things within a module that have a leading _ are intended to be private with regard to anything other than a unit test.